docs/celestia-architecture/adr-009-info-reserved-byte.md

```diff
@@ -0,0 +1,120 @@
+# ADR 009: InfoReservedByte
+
+## Changelog
+
+- 2022/9/22: inital draft
+
+## Context
+
+The current message share format is as follows:
+
+- `nid (8 bytes) | message length (varint) | share data` for shares at the start of a message.
+- `nid (8 bytes) | share data` for other shares.
+
+The current message share format poses two challenges:
+
+1. It is difficult to make changes to the message share format in a backwards compatible way because clients can't determine which version of the message share format an individual share conforms to.
+1. It is not possible for a client that samples a random message share to determine if the share is the start of a message or a contiguous share for a multi-share message.
+
+## Proposal
+
+- `nid (8 bytes) | info byte | message length (varint) | share data` for shares at the start of a message.
+- `nid (8 bytes) | info byte | share data` for other shares.
+
+Where info byte is a byte with the following structure:
+
+- the first 7 bits are reserved for the version information in big endian form (initially, this will just be 0000000 until further notice);
+- the last bit is a "message start indicator", that is 1 if the share is at the start of a message.
+
+Rationale:
+
+1. The message start indicators allow clients to parse a whole message in the middle of a namespace, without needing to read the whole namespace.
+1. The version bits allow us to upgrade the share format in the future, if we need to do so in such a way that different share formats can be mixed within a block.
+
+## Alternative Approaches
+
+// TODO
+
+## Decision
+
+// TODO
+
+## Implementation Details
+
+### Protobuf
+
+1. Add `Version` to [`MsgWirePayForData`](https://github.com/celestiaorg/celestia-app/blob/main/proto/payment/tx.proto#L19)
+1. Add `Version` to [`MsgPayForData`](https://github.com/celestiaorg/celestia-app/blob/main/proto/payment/tx.proto#L44)
+
+**NOTE**: Protobuf does not support the byte type (see [Scalar Value Types](https://developers.google.com/protocol-buffers/docs/proto3#scalar)) so a `uint32` will be used for `Version`. Since `Version` is constrained to 2^7 bits (0 to 127 inclusive), a `Version` outside the supported range (i.e. 128) will seriealize / deserialize correctly but be considered invalid by the application.
```

Comment:
I not sure if its worth putting specifically in the ADR, but for posterity/other reviewers, what the user signs over will only be a single byte + protobuf overhead, as protobuf uses a varint encoding for all number types besides floats and what not.

Reply:
Thanks for sharing!

> as protobuf uses a varint encoding for all number types

Agreed based on https://developers.google.com/protocol-buffers/docs/encoding which states Varint is used for `uint32`.  It's not clear to me how to estimate protobuf overhead so added a sentence in https://github.com/celestiaorg/celestia-core/pull/843/commits/46203433b44d50bbcfc1e6df649da3ffe4a53c76